# Houseing Price Prediction with Decision Tree and Random Forest

This is an introductory ML project from [Kaggle's Micro Courses.](https://www.kaggle.com/learn/intro-to-machine-learning)

In this project, I used Random Forest and Decision Trees to predict the housing prices in Melbourne, Australia.

In the end, I compared the two model using MAE(Mean Absolute Error).

## 1. Importing Pandas

In [1]:
import pandas as pd

## 2. Loading Data

In [2]:
path = 'melb_data.csv'
melb_data = pd.read_csv(path)

In [3]:
# This is How the data looks like!
melb_data.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [4]:
# A brief summary of the data
melb_data.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

## 3. Droping rows with any missing values

In [5]:
melb_data = melb_data.dropna(axis=0)

In [6]:
# After the data cleansing
melb_data.describe()

Rooms         Price     Distance     Postcode     Bedroom2  \
count  6196.000000  6.196000e+03  6196.000000  6196.000000  6196.000000   
mean      2.931407  1.068828e+06     9.751097  3101.947708     2.902034   
std       0.971079  6.751564e+05     5.612065    86.421604     0.970055   
min       1.000000  1.310000e+05     0.000000  3000.000000     0.000000   
25%       2.000000  6.200000e+05     5.900000  3044.000000     2.000000   
50%       3.000000  8.800000e+05     9.000000  3081.000000     3.000000   
75%       4.000000  1.325000e+06    12.400000  3147.000000     3.000000   
max       8.000000  9.000000e+06    47.400000  3977.000000     9.000000   

          Bathroom          Car      Landsize  BuildingArea    YearBuilt  \
count  6196.000000  6196.000000   6196.000000   6196.000000  6196.000000   
mean      1.576340     1.573596    471.006940    141.568645  1964.081988   
std       0.711362     0.929947    897.449881     90.834824    38.105673   
min       1.000000     0.000000      0.000000      0.000000  1196.000000   
25%       1.000000     1.000000    152.000000     91.000000  1940.000000   
50%       1.000000     1.000000    373.000000    124.000000  1970.000000   
75%       2.000000     2.000000    628.000000    170.000000  2000.000000   
max       8.000000    10.000000  37000.000000   3112.000000  2018.000000   

         Lattitude   Longtitude  Propertycount  
count  6196.000000  6196.000000    6196.000000  
mean    -37.807904   144.990201    7435.489509  
std       0.075850     0.099165    4337.698917  
min     -38.164920   144.542370     389.000000  
25%     -37.855438   144.926198    4383.750000  
50%     -37.802250   144.995800    6567.000000  
75%     -37.758200   145.052700   10175.000000  
max     -37.457090   145.526350   21650.000000

## 4. Splitting the Data into Training Sets and Validation Sets

In [7]:
from sklearn.model_selection import train_test_split
y = melb_data.Price # Prediction Target
features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
            'YearBuilt', 'Lattitude', 'Longtitude']
X = melb_data[features] # Features

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

## 5. Build, Train, and Predict

In [8]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error

# Since we can specify the depth of the tree, going to define a function to compare depths of the model, using MAE
# For this model, we will test by using 5, 50, 500, and 5000 maximum leaf nodes.
def decision_tree_best_mae(train_X, val_X, train_y, val_y):
    min_mae = 0
    for num in [5, 50, 500, 5000]:
        # 5-1 Decision Tree
        model = DecisionTreeRegressor(max_leaf_nodes = num, random_state=0)
        model.fit(train_X, train_y) # Train our model
        preds_val = model.predict(val_X) # Make a prediction
        mae = mean_absolute_error(val_y, preds_val)
        if min_mae == 0:
            min_mae = mae
        elif min_mae >= mae:
            min_mae = mae
    return(min_mae)

In [9]:
# 5-2 Random Forest
from sklearn.ensemble import RandomForestRegressor

forest_model = RandomForestRegressor(random_state = 0)
forest_model.fit(train_X, train_y) # Train our model
forest_preds = forest_model.predict(val_X) # Make a prediction.

## 6. Compare the Models!

In [10]:
print('Random Forest MAE: %d' % (mean_absolute_error(val_y, forest_preds)))
print('Decision Tree MAE: %d' % (decision_tree_best_mae(train_X, val_X, train_y, val_y)))

Random Forest MAE: 193528
Decision Tree MAE: 243495


## 7. Conclusion

As it is noticeable by the name MAE, MAE is the number is the average of difference between the actual values of data and predictions.

| | Decision Tree | Random Forest |
| --- | :--- | :--- |
| MAE | 243495 | 193528 |


As you can see by the result, Random Forest showed way better performance than our Decision Tree.
It makes sense because Random Forest is a machine that uses multiple Decision Trees to average their outcome.

This is the end of this project.